// @hashgraph/sdk/lib/schedule/ScheduleId.cjs
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var entity_id = _interopRequireWildcard(require("../EntityIdHelper.cjs"));

var proto = _interopRequireWildcard(require("@hashgraph/proto"));

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

/**
 * @typedef {import("long").Long} Long
 */

/**
 *
 * @augments {EntityId<proto.IScheduleID>}
 */
class ScheduleId {
  /**
   * @param {number | Long | import("../EntityIdHelper.js").IEntityId} props
   * @param {(number | Long)=} realm
   * @param {(number | Long)=} num
   */
  constructor(props, realm, num) {
    const [shard_num, realm_num, schedule_num] = entity_id.constructor(props, realm, num);
    this.shard = shard_num;
    this.realm = realm_num;
    this.num = schedule_num;
  }
  /**
   * @param {string} text
   * @returns {ScheduleId}
   */


  static fromString(text) {
    return new ScheduleId(...entity_id.fromString(text));
  }
  /**
   * @internal
   * @param {proto.IScheduleID} id
   * @returns {ScheduleId}
   */


  static _fromProtobuf(id) {
    return new ScheduleId({
      shard: id.shardNum != null ? id.shardNum : 0,
      realm: id.realmNum != null ? id.realmNum : 0,
      num: id.scheduleNum != null ? id.scheduleNum : 0
    });
  }
  /**
   * @param {Uint8Array} bytes
   * @returns {ScheduleId}
   */


  static fromBytes(bytes) {
    return ScheduleId._fromProtobuf(proto.ScheduleID.decode(bytes));
  }
  /**
   * @param {string} address
   * @returns {ScheduleId}
   */


  static fromSolidityAddress(address) {
    return new ScheduleId(...entity_id.fromSolidityAddress(address));
  }
  /**
   * @internal
   * @override
   * @returns {proto.ScheduleID}
   */


  _toProtobuf() {
    return {
      scheduleNum: this.num,
      shardNum: this.shard,
      realmNum: this.realm
    };
  }
  /**
   * @override
   * @returns {string}
   */


  toString() {
    return `${this.shard.toString()}.${this.realm.toString()}.${this.num.toString()}`;
  }
  /**
   * @returns {Uint8Array}
   */


  toBytes() {
    return proto.ScheduleID.encode(this._toProtobuf()).finish();
  }

}

exports.default = ScheduleId;
