// @hashgraph/sdk/lib/topic/TopicId.cjs
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var entity_id = _interopRequireWildcard(require("../EntityIdHelper.cjs"));

var _proto = require("@hashgraph/proto");

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

/**
 * @typedef {import("long").Long} Long
 */

/**
 * @namespace proto
 * @typedef {import("@hashgraph/proto").ITopicID} proto.ITopicID
 */

/**
 * Unique identifier for a topic (used by the consensus service).
 */
class TopicId {
  /**
   * @param {number | Long | import("../EntityIdHelper.js").IEntityId} props
   * @param {(number | Long)=} realm
   * @param {(number | Long)=} num
   */
  constructor(props, realm, num) {
    const [shard_num, realm_num, topic_num] = entity_id.constructor(props, realm, num);
    this.shard = shard_num;
    this.realm = realm_num;
    this.num = topic_num;
  }
  /**
   * @param {string} text
   * @returns {TopicId}
   */


  static fromString(text) {
    return new TopicId(...entity_id.fromString(text));
  }
  /**
   * @param {proto.ITopicID} id
   * @returns {TopicId}
   */


  static _fromProtobuf(id) {
    return new TopicId({
      shard: id.shardNum != null ? id.shardNum : 0,
      realm: id.realmNum != null ? id.realmNum : 0,
      num: id.topicNum != null ? id.topicNum : 0
    });
  }
  /**
   * @param {Uint8Array} bytes
   * @returns {TopicId}
   */


  static fromBytes(bytes) {
    return TopicId._fromProtobuf(_proto.TopicID.decode(bytes));
  }
  /**
   * @override
   * @returns {proto.ITopicID}
   */


  _toProtobuf() {
    return {
      topicNum: this.num,
      shardNum: this.shard,
      realmNum: this.realm
    };
  }
  /**
   * @override
   * @returns {string}
   */


  toString() {
    return `${this.shard.toString()}.${this.realm.toString()}.${this.num.toString()}`;
  }
  /**
   * @returns {Uint8Array}
   */


  toBytes() {
    return _proto.TopicID.encode(this._toProtobuf()).finish();
  }

}

exports.default = TopicId;
